Publish A New Extension: Parrot-Translate

Haojen opened the pull request:
## Description

A Powerful and Easy to use translation, Support TTS and lowerCamelCase/ALL_UPPERCASE Copy mode, More features introduce can visit this project readme https://github.com/Haojen/raycast-Parrot

## Screencast

### Normal translate 
<img width="839" alt="Screen Shot 2022-03-04 at 12 48 13 PM" src="https://user-images.githubusercontent.com/6928850/156701295-84fad634-1136-4d80-8d47-5b693b8337c3.png">

### lowerCaseCopy translate result
<img width="825" alt="lowerCopyMode" src="https://user-images.githubusercontent.com/6928850/156700677-79a58a89-c1d1-47d2-8d24-4f64b6415157.png">

  ### UPPERCASE_COPY translate result
<img width="813" alt="UppercastCopy" src="https://user-images.githubusercontent.com/6928850/156700689-85b7b51d-810c-4502-9b89-8e8409319869.png">

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed in the `media` folder

pernielsentikaer commented on extensions/parrot-translate/src/parrot.tsx:
```diff
@@ -0,0 +1,158 @@
+            isLoading={isLoadingState}
+            searchBarPlaceholder={"Translate to.."}
+            onSearchTextChange={(inputText) => onInputChangeEvt(inputText)}
+            navigationTitle={
+                translateFromLanguageState?.languageTitle && currentTargetLanguageState?.languageTitle
+                    ? `Parrot: ${translateFromLanguageState.languageTitle} → ${currentTargetLanguageState.languageTitle}`
+                    : "Parrot"
+            }
```
```suggestion
```
The [navigationTitle](https://developers.raycast.com/basics/prepare-an-extension-for-store#navigation-title) should not be set here in the root

pernielsentikaer commented on extensions/parrot-translate/src/components.tsx:
```diff
@@ -0,0 +1,141 @@
+                    <Action.CopyToClipboard
+                        onCopy={() => preferences.isAutomaticPaste && Clipboard.paste(textItem.value)}
+                        shortcut={{ modifiers: ["cmd"], key: shortcutKeyEquivalent[key] }}
+                        title={`Copy ${title}`}
```
```suggestion
                        title="Copy Result"
```

pernielsentikaer replied:
You need to change this too :)

pernielsentikaer reviewed (changes requested):
Hi @Haojen 👋 

Thanks for your contribution 💯 

I have some feedback and suggestions ready for you 💪 

- When opening the extension, then there is a button saying “Chinese-Simplified” – if I click the button and search it, it's then searching for “Autodetect-Chinese” – EDIT: Now I can see what it does, maybe it should first be visible after search – kind of confusing as it is right now 😄 

- Missing an option to have autodetect in preferences for the “First language”

- Instead of “Automatic Paste” in preferences, then you should add [Action.Paste](https://developers.raycast.com/api-reference/user-interface/actions#action.paste) instead 

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

pernielsentikaer commented:
Hi @Haojen 👋 

Please remember that one PR is. only for one extension, could you remove it from this PR?

On a sidenote, did you see that we already have a [speedtest extension](https://www.raycast.com/tonka3000/speedtest) in the store?

Haojen commented:
@pernielsentikaer  yeah, sorry about that. I developer another Extension, I forget this PR does not merge. I will revert. I searched and checked, but I really don't see this speedtest extension.  so so so sad.

Haojen commented:
@pernielsentikaer I 13 days ago. this too fast happened, I can't always check has the too same extension. :(

pernielsentikaer commented on extensions/parrot-translate/src/parrot.tsx:
```diff
@@ -0,0 +1,153 @@
+    return (
+        <List
+            isLoading={isLoadingState}
+            searchBarPlaceholder={"Translate to"}
```
```suggestion
            searchBarPlaceholder={"Translate to"}
            throttle={true}
```

Haojen replied:
The throttle API does not a good choice. here is why: https://github.com/raycast/extensions/issues/766

pernielsentikaer commented on extensions/parrot-translate/src/parrot.tsx:
```diff
@@ -0,0 +1,153 @@
+    useEffect(() => {
+        if (!inputState) return // Prevent when mounted run
+
+        clearTimeout(delayFetchTranslateAPITimer)
+        delayFetchTranslateAPITimer = setTimeout(() => {
+            requestYoudaoAPI(inputState, translateTargetLanguage.languageId).then((res) => {
+                const resData: ITranslateResult = res.data
+
+                const [a, b] = resData.l.split("2") // en2zh
+
+                if (a === b) {
+                    updateTranslateTargetLanguage(a === preferences.lang1 ? defaultLanguage2 : defaultLanguage1)
+                    return
+                }
+
+                updateLoadingState(false)
+                fetchResultStateCode = res.data.errorCode
+                updateTranslateResultState(reformatTranslateResult(resData))
+
+                updateTranslateFromLanguageState(getItemFromLanguageList(a))
+                updateCurrentTargetLanguageState(getItemFromLanguageList(b))
+            })
+        }, 900)
```
```suggestion
        requestYoudaoAPI(inputState, translateTargetLanguage.languageId).then((res) => {
            const resData: ITranslateResult = res.data
    
            const [a, b] = resData.l.split("2") // en2zh
    
            if (a === b) {
                updateTranslateTargetLanguage(a === preferences.lang1 ? defaultLanguage2 : defaultLanguage1)
                return
            }
    
            updateLoadingState(false)
            fetchResultStateCode = res.data.errorCode
            updateTranslateResultState(reformatTranslateResult(resData))
    
            updateTranslateFromLanguageState(getItemFromLanguageList(a))
            updateCurrentTargetLanguageState(getItemFromLanguageList(b))
        })
```
With throttle on the list, this shouldn't be necessary

pernielsentikaer commented on extensions/parrot-translate/src/components.tsx:
```diff
@@ -0,0 +1,145 @@
+    return (
+        <ActionPanel.Section>
+            {finalTextArray.map((textItem, key) => {
+                const title = changeTextCopyStyle(textItem.title)
```
```suggestion
```
Redundant

Haojen replied:
the title key content and value key content are not the same when the translated resulting length is too long. so I need to keep the title and value key.

Haojen replied:
<img width="796" alt="image" src="https://user-images.githubusercontent.com/6928850/158006003-22cb5e43-a4ca-4b78-a799-809e1397b735.png">

The Title is: `Copy All`, The value key actually is `How are you；How Do You Do；Are You OK`

pernielsentikaer replied:
Ah, fair!

pernielsentikaer reviewed (changes requested):
Hi @Haojen 👋 

I checked the latest update, and I have some feedback for you to address 😄 

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

pernielsentikaer commented:
Hi @Haojen 👋 

There is a problem now that throttle is not used, see below - maybe you can find a way to have them both, as in my last video? Just added throttle:

Before:

https://user-images.githubusercontent.com/38033107/158200354-6cb9eac5-8fa3-4feb-9a8c-5a7508c93621.mp4

After:

https://user-images.githubusercontent.com/38033107/158200383-088c0168-4897-4a39-a48f-dfc34f9bc7c7.mp4

pernielsentikaer reviewed (changes requested):
Hi @Haojen 👋 

Sorry for the longer response time 😄 

I tried to start all over and installing it again, it selected this:

![wTGyO4TOarZj6Dxn](https://user-images.githubusercontent.com/38033107/158519975-50b8d959-1977-47cb-9d62-2b6d830ff257.png)

However, I end up with this result:

![JVNQriVtkYRj4CXo](https://user-images.githubusercontent.com/38033107/158520024-9aeabeb9-4c65-4b6e-aaa8-bbc0d25add2a.png)

It's the same when I change to another language, it seems like it's stuck with Chinese somehow.

Regarding the flicker on the details button, I would like it to be fixed before we can merge also, it's not giving a good UX

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

Haojen commented:
> Hi @Haojen 👋
> 
> Sorry for the longer response time 😄
> 
> I tried to start all over and installing it again, it selected this:
> 
> ![wTGyO4TOarZj6Dxn](https://user-images.githubusercontent.com/38033107/158519975-50b8d959-1977-47cb-9d62-2b6d830ff257.png)
> 
> However, I end up with this result:
> 
> ![JVNQriVtkYRj4CXo](https://user-images.githubusercontent.com/38033107/158520024-9aeabeb9-4c65-4b6e-aaa8-bbc0d25add2a.png)
> 
> It's the same when I change to another language, it seems like it's stuck with Chinese somehow.
> 
> Regarding the flicker on the details button, I would like it to be fixed before we can merge also, it's not giving a good UX
> 
> Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

1. Fixed always is Chinese language bug
2. Fixed action flash bug
3. Improved translate speed
4. Improved stability and detail optimization

pernielsentikaer reviewed (changes requested):
Hi @Haojen 👋 

Thanks for the update, it does indeed work better now 🔥 

We're now adding support for [categories](https://developers.raycast.com/basics/prepare-an-extension-for-store#categories), [screenshots](https://developers.raycast.com/basics/prepare-an-extension-for-store#screenshots) and [changelog](https://developers.raycast.com/basics/prepare-an-extension-for-store#version-history) – do you have time for adding this too before final review?

Haojen commented:
@pernielsentikaer Good to hear, I later I'll add it later

Haojen commented:
@pernielsentikaer  I added categories, screenshots and changelog.  waiting so long : )

pernielsentikaer commented on extensions/parrot-translate/package.json:
```diff
@@ -0,0 +1,292 @@
+        }
+    ],
+    "dependencies": {
+        "@raycast/api": "^1.30.0",
```
```suggestion
        "@raycast/api": "^1.31.0",
```

pernielsentikaer commented on extensions/parrot-translate/src/parrot.tsx:
```diff
@@ -0,0 +1,178 @@
+    return (
+        <List
+            isLoading={isLoadingState}
+            searchBarPlaceholder={"Translate to"}
```
```suggestion
            searchBarPlaceholder={"Translate text"}
```

pernielsentikaer commented on extensions/parrot-translate/src/parrot.tsx:
```diff
@@ -0,0 +1,178 @@
+                </ActionPanel>
+            }
+        >
+            <ListDetail />
```
```suggestion
            <List.EmptyView
                icon="parrot-icon-small.png"
                title="Type something to get started"
            />
            <ListDetail />
```
Add the attached icon to assets 😄

pernielsentikaer reviewed (changes requested):
Hi @Haojen 👋 

Thanks for the update, much appreciated 🙏 

I think we're about to be ready, I have the last feedback for you before merging 💪


This is the image for assets folder:
![parrot-icon-small](https://user-images.githubusercontent.com/38033107/161850271-6534a423-6d59-44f7-a8b2-79d1ac8c782c.png) 

Request a new review when you are ready, feel free to contact me here or at [Slack](https://raycastcommunity.slack.com/) if you have any questions.

pernielsentikaer reviewed (approved):
Hi @Haojen 👋 

Awesome, let's get the first version up and running and do a version 2 with the rest as we spoke about 💪
